Modules/Scripted/VectorToScalarVolume/VectorToScalarVolume.py

```diff
@@ -26,82 +63,420 @@ def __init__(self, parent):
+    self.parameterNodeSelector.showHidden = True
+    self.parameterNodeSelector.showChildNodeTypes = False
+    self.parameterNodeSelector.baseName = "VectorToScalarVolume"
+    self.parameterNodeSelector.setMRMLScene(slicer.mrmlScene)
+    self.parameterNodeSelector.toolTip = "Pick parameter set"
+    parametersFormLayout.addRow("Parameter set: ", self.parameterNodeSelector)
 
     # Add vertical spacer
     self.layout.addStretch(1)
 
+    # Connections
+    self.parameterNodeSelector.connect('currentNodeChanged(vtkMRMLNode*)', self.setParameterNode)
+
+    # updateParameterNodeFromGui
+    self.inputSelector.connect("currentNodeChanged(vtkMRMLNode*)", self.updateParameterNodeFromGui)
+    self.outputSelector.connect("currentNodeChanged(vtkMRMLNode*)", self.updateParameterNodeFromGui)
+    self.conversionMethodWidget.methodSelectorComboBox.connect('currentIndexChanged(int)', self.updateParameterNodeFromGui)
+    self.conversionMethodWidget.componentsComboBox.connect('currentIndexChanged(int)', self.updateParameterNodeFromGui)
+
+    self.applyButton.connect('clicked(bool)', self.onApply)
+
+    self.setDefaultParameters()
+
+  def cleanup(self):
+    self.removeObservers()
+
+  def setDefaultParameters(self):
+    """Set computation parameters to their default value"""
+    self.conversionMethodWidget.setConversionMethod(VectorToScalarVolumeLogic.LUMINANCE)
+    self.conversionMethodWidget.componentsComboBox.insertItem(self.conversionMethodWidget.EXTRACT_COMPONENT_NONE, "None")
+
+
+  def parameterNode(self):
+    return self._parameterNode
+
+  def setParameterNode(self, inputParameterNode):
+    if inputParameterNode == self._parameterNode:
+      return
+    if self._parameterNode is not None:
+      self.removeObserver(self._parameterNode, vtk.vtkCommand.ModifiedEvent, self.updateGuiFromMRML)
+    if inputParameterNode is not None:
+      self.addObserver(inputParameterNode, vtk.vtkCommand.ModifiedEvent, self.updateGuiFromMRML)
+    self._parameterNode = inputParameterNode
+
+  def inputVolumeNode(self):
+    return self.inputSelector.currentNode()
+
+  def setInputVolumeNode(self, node):
+    if isinstance(node, basestring):
+      node = getNode(node)
+    self.inputSelector.setCurrentNode(node)
+    self.conversionMethodWidget.setInputVolumeNode(node)
+
+  def outputVolumeNode(self):
+    return self.outputSelector.currentNode()
+
+  def setOutputVolumeNode(self, node):
+    if isinstance(node, basestring):
+      node = getNode(node)
+    self.outputSelector.setCurrentNode(node)
+
+  def updateGuiFromMRML(self, caller=None, event=None):
+    """
+    Query all the parameters in the parameterNode,
+    and update the GUI state accordingly if something has changed.
+    """
+    self.applyButton.enabled = all([self.inputVolumeNode(), self.outputVolumeNode(), self.parameterNode()])
+    if self.parameterNode is None:
+      return
+    self.setInputVolumeNode(self._parameterNode.GetParameter("InputVectorVolume"))
```

Comment:
need to call blocksignals for each widget when modifying from parameter node to make sure updateMRMLFromGUI method is not called in the middle of updateGUIFromMRML.

Reply:
Since the widget is setting the value of the parameter node to what it is already, no node modified event should be invoked.

As a precautionary measure, I suggest we introduce a context managed named like `BlockedObjectSignal`

Reply:
I haven't block anything, but created a context manager to block signals in a list of qobjects.